itests/mir_test.go

```diff
@@ -1151,6 +1151,117 @@ func TestMirBasic_AllNodesMiningWithMessaging(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	g, ctx := errgroup.WithContext(ctx)
+
+	defer func() {
+		t.Logf("[*] defer: cancelling %s context", t.Name())
+		cancel()
+		err := g.Wait()
+		require.NoError(t, err)
+		t.Logf("[*] defer: system %s stopped", t.Name())
+	}()
+
+	nodes, validators, ens := kit.EnsembleWithMirValidators(t, MirTotalValidatorNumber)
+	ens.InterconnectFullNodes().BeginMirMiningWithTestAndConsensusConfigs(ctx, g, validators,
+		kit.DefaultMirTestConfig(),
+		&mir.ConsensusConfig{
+			SegmentLength:                1,
+			ConfigOffset:                 2,
+			MaxProposeDelay:              5 * time.Second,
+			MaxTransactionsInBatch:       1024,
+			PBFTViewChangeSegmentTimeout: 6 * time.Second,
+			PBFTViewChangeSNTimeout:      6 * time.Second,
+		},
+	)
+
+	err := kit.AdvanceChain(ctx, TestedBlockNumber, nodes...)
+	require.NoError(t, err)
+	from, err := nodes[0].IsSyncedWith(ctx, 0, nodes[1:]...)
+	require.NoError(t, err)
+
+	src, err := nodes[0].WalletDefaultAddress(ctx)
+	require.NoError(t, err)
+
+	dst, err := nodes[1].WalletDefaultAddress(ctx)
+	require.NoError(t, err)
+
+	// Send N messages from node 0 to node 1.
+	n := 30
+	var cids []cid.Cid
+	for i := 0; i < n; i++ {
+		msg, err := nodes[0].MpoolPushMessage(ctx, &types.Message{
+			From:  src,
+			To:    dst,
+			Value: big.Zero(),
+		}, nil)
+		require.NoError(t, err)
+		cids = append(cids, msg.Cid())
+	}
+	require.Equal(t, n, len(cids))
+
+	err = kit.AdvanceChain(ctx, TestedBlockNumber, nodes...)
+	require.NoError(t, err)
+	_, err = nodes[0].IsSyncedWith(ctx, from, nodes[1:]...)
+	require.NoError(t, err)
+
+	for _, id := range cids {
+		err = kit.MirNodesWaitForMsg(ctx, id, nodes[0])
+		require.NoError(t, err)
+	}
+
+	var maxHeight abi.ChainEpoch
+	for i, msgCID := range cids {
+		searchRes, err := nodes[0].StateSearchMsg(ctx, types.EmptyTSK, msgCID, api.LookbackNoLimit, true)
+		require.NoError(t, err)
+
+		if searchRes != nil &&
+			searchRes.Height > maxHeight {
+			maxHeight = searchRes.Height
+		}
+
+		t.Logf(">>> message number %d with CID %v is in the block %d", i, msgCID, searchRes.Height)
+	}
+	t.Logf(">>> target epoch is %d", maxHeight)
+
+	t.Logf(">>> checking nonces")
+
+	for i := 0; i <= int(maxHeight); i++ {
+		ts, err := nodes[0].ChainGetTipSetByHeight(ctx, abi.ChainEpoch(i), types.TipSetKey{})
+		require.NoError(t, err)
+
+		for _, b := range ts.Blocks() {
+
+			bh, err := nodes[0].ChainGetBlock(ctx, b.Cid())
+			require.NoError(t, err)
+
+			msgs, err := nodes[0].ChainGetBlockMessages(ctx, bh.Cid())
+			require.NoError(t, err)
+
+			t.Logf(">>> there are %d Secpk messages in the block %v", len(msgs.SecpkMessages), i)
+			t.Logf(">>> there are %d BLS messages in the block %v", len(msgs.BlsMessages), i)
+
+			var lastNonce uint64
+
+			for _, msg := range msgs.BlsMessages {
+				t.Logf(">>> BLS message nonce is %d", msg.Nonce)
+				require.GreaterOrEqual(t, msg.Nonce, lastNonce)
```

Comment:
Why equal? Shouldn't it be strictly greater? Even more, this should check that is the subsequent nonce, right?

Reply:
Great gatch!